heron/stmgr/src/cpp/manager/instance-server.cpp

```diff
@@ -74,7 +74,10 @@ const sp_string METRIC_TIME_SPENT_BACK_PRESSURE_AGGR = "__server/__time_spent_ba
 const sp_string METRIC_TIME_SPENT_BACK_PRESSURE_COMPID = "__time_spent_back_pressure_by_compid/";
 // Prefix for connection buffer's metrics
 const sp_string CONNECTION_BUFFER_BY_INSTANCEID = "__connection_buffer_by_instanceid/";
-// Prefix for connection buffer's length metrics
+// Prefix for connection buffer's length metrics. This is different
+// from METRIC_DATA_TUPLES_TO_INSTANCES as that counts
+// the tuples when they are sent to the instance -- this metric
+// will be used to count the tuples as they are received
```

Comment:
Why do we need such a separate metric to count tuples when they are received?  In effect, the sent tuples should be equal to received tuples. If the existing one can server the purpose, I suggest to use it directly instead of adding new metrics.

Reply:
Would the two not be different in case of backpressure?

Reply:
In the case of backpressure, only the spouts stop reading from the external data source(e.g, kafka stream). 

All the internal tuples are still sent and processed. 

Could you elaborate why they are different in such case?

Reply:
From what I understand, tuples can be dropped when they are being removed from the queue because of backpressure.
https://github.com/apache/incubator-heron/blob/571a90b7c8452f7922cdc47b91e52de66f5edd70/heron/stmgr/src/cpp/manager/instance-server.cpp#L432
In addition, I'd like to measure arrival rate of tuples at the stream manager and I think that because tuples are only drained when a certain number of tuples have arrived, that may not translate accurately?